crates/holochain_p2p/src/op_store.rs

```diff
@@ -0,0 +1,403 @@
+use crate::event::{HolochainP2pEvent, HolochainP2pEventSender};
+use bytes::Bytes;
+use futures::future::BoxFuture;
+use ghost_actor::GhostSender;
+use holo_hash::{DhtOpHash, DnaHash};
+use holochain_serialized_bytes::prelude::decode;
+use holochain_sqlite::db::{DbKindDht, DbWrite};
+use holochain_sqlite::rusqlite::types::Value;
+use holochain_sqlite::sql::sql_dht::{
+    OPS_BY_ID, OP_HASHES_IN_TIME_SLICE, OP_HASHES_SINCE_TIME_BATCH,
+};
+use holochain_state::prelude::{named_params, StateMutationResult};
+use holochain_types::dht_op::DhtOpHashed;
+use holochain_types::prelude::DhtOp;
+use kitsune2_api::*;
+use std::collections::HashSet;
+use std::fmt::{Debug, Formatter};
+use std::rc::Rc;
+
+/// Holochain implementation of the Kitsune2 [OpStore].
+pub struct HolochainOpStore {
+    db: DbWrite<DbKindDht>,
+    dna_hash: DnaHash,
+    sender: GhostSender<HolochainP2pEvent>,
+}
+
+impl Debug for HolochainOpStore {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("HolochainOpStore")
+            .field("db", &self.db)
+            .finish()
+    }
+}
+
+impl HolochainOpStore {
+    /// Create a new [HolochainOpStore].
+    pub fn new(
+        db: DbWrite<DbKindDht>,
+        dna_hash: DnaHash,
+        sender: GhostSender<HolochainP2pEvent>,
+    ) -> HolochainOpStore {
+        Self {
+            db,
+            dna_hash,
+            sender,
+        }
+    }
+}
+
+impl OpStore for HolochainOpStore {
+    fn process_incoming_ops(&self, op_list: Vec<Bytes>) -> BoxFut<'_, K2Result<Vec<OpId>>> {
+        Box::pin(async move {
+            let dht_ops = op_list
+                .into_iter()
+                // Filter to make casting the size below safe
+                .filter(|op| op.len() <= u32::MAX as usize)
+                .map(|op| {
+                    Ok((
+                        op.len() as u32,
+                        decode::<_, DhtOp>(op.as_ref())
+                            .map_err(|e| K2Error::other_src("Could not decode op", e))?,
+                    ))
+                })
+                .collect::<K2Result<Vec<(u32, DhtOp)>>>()?;
+
+            let ids = dht_ops
+                .iter()
+                .map(|(_, op)| {
+                    let op_hashed = DhtOpHashed::from_content_sync(op.clone());
+                    OpId::from(Bytes::copy_from_slice(op_hashed.hash.get_raw_36()))
```

Comment:
I'm okay if we do this later, but seems like we might want to add a From impl on the hash type directly.

Reply:
I feel like a kitsune2 feature that adds it to holohash is the way to go, but I'm totally fine with an extension trait in holochain_p2p as well.